File: Clockio/app/src/main/java/com/natasha/clockio/base/util/RetrofitInterceptor.java
package com.natasha.clockio.base.util;

import android.text.TextUtils;
import android.util.Log;
import okhttp3.Credentials;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.IOException;

//https://jakewharton.com/making-retrofit-work-for-you/
@Singleton
public class RetrofitInterceptor implements Interceptor {
    private static final String TAG = RetrofitInterceptor.class.getSimpleName();

    private String token = "";
    private String clientId, clientSecret;

    @Inject public RetrofitInterceptor() {
    }

    public void setToken(String token) {
        this.token = token;
    }

    public void setBasic(String clientId, String clientSecret) {
        this.clientId = clientId;
        this.clientSecret = clientSecret;
    }

    @NotNull
    @Override
    public Response intercept(@NotNull Chain chain) throws IOException {
        Request request = chain.request();
        if (request.url().encodedPath().equals("/oauth/token")) {
            request = request.newBuilder()
                    .header("Authorization", Credentials.basic(clientId, clientSecret)).build();
            Log.d(TAG, "retrofit add header basic");
        } else if (!TextUtils.isEmpty(token)){
            Log.d(TAG, "retrofit add header token");
            request = request.newBuilder()
                    .addHeader("Authorization", "Bearer "  +token).build();
        } else {
            Log.d(TAG, "retrofit add header NOTHING!");
        }
        return chain.proceed(request);
    }
}

File: Clockio/app/src/main/java/com/natasha/clockio/base/util/RetrofitInterceptor.kt
package com.natasha.clockio.base.util

import android.content.SharedPreferences
import android.text.TextUtils
import android.util.Log
import com.natasha.clockio.base.constant.PreferenceConst
import okhttp3.Credentials
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response

import javax.inject.Inject
import javax.inject.Singleton
import java.io.IOException

//https://jakewharton.com/making-retrofit-work-for-you/
@Singleton
class RetrofitInterceptor @Inject constructor(private val sharedPref: SharedPreferences) : Interceptor {

    private var token = ""
    private var clientId: String? = null
    private var clientSecret: String? = null
    private var existToken = sharedPref.getString(PreferenceConst.ACCESS_TOKEN_KEY, "")

    fun setToken(token: String) {
        this.token = token
    }

    fun setBasic(clientId: String, clientSecret: String) {
        this.clientId = clientId
        this.clientSecret = clientSecret
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        var request = chain.request()
        if (request.url.encodedPath == "/oauth/token") {
            val basic = Credentials.basic(clientId!!, clientSecret!!)
            request = request.newBuilder()
                .addHeader("Authorization", basic)
                .addHeader("Content-Type", "application/json")
                .build()
            Log.d(TAG, "retrofit add header basic $basic")
        } else if (!TextUtils.isEmpty(existToken)) {
            Log.d(TAG, "retrofit add header token from Shared Pref $existToken")
            request = request.newBuilder()
                .addHeader("Authorization", "Bearer $existToken")
                .addHeader("Content-Type", "application/json")
                .build()
        } else if (!TextUtils.isEmpty(token)) {
            Log.d(TAG, "retrofit add header token from Login $token")
            request = request.newBuilder()
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .build()
        } else {
            Log.d(TAG, "retrofit add header NOTHING!")
        }
        return chain.proceed(request)
    }

    companion object {
        private val TAG = RetrofitInterceptor::class.java.simpleName
    }
}
